### PR TITLE
Add Semantic Kernel sample projects from build 2024

### PR DESCRIPTION
**Pull request by zewditu**
This PR takes the code from the Build 2024  demo [here](https://www.youtube.com/watch?v=jrNfKeGSuCg&t=179s) and breaks it up into incremental changes.

Each project is self contained and can be run on it's own to learn a concept.

### CODE REVIEW

File: src/build-2024/01 - Hello Semantic Kernel/README.md [hunk 1]
```diff
@@ -0,0 +1,37 @@
+﻿This project demonstrates a simple console application using the Microsoft Semantic Kernel to interact with the OpenAI GPT-3.5-turbo model. The application allows users to have a stateless chat with the AI, meaning the AI does not retain memory of previous messages in the conversation.
+
+## Overview
+The main code utilizes the Microsoft Semantic Kernel to integrate with the OpenAI GPT-3.5-turbo model. Here's a breakdown of what the code does:
+1.Import the Microsoft Semantic Kernel:
```
Comment (stephentoub) [inline src/build-2024/01 - Hello Semantic Kernel/README.md:5]:
```suggestion
1.Import the core namespace for the Semantic Kernel library:
```

File: src/build-2024/01 - Hello Semantic Kernel/README.md [hunk 2]
```diff
@@ -0,0 +1,37 @@
+﻿This project demonstrates a simple console application using the Microsoft Semantic Kernel to interact with the OpenAI GPT-3.5-turbo model. The application allows users to have a stateless chat with the AI, meaning the AI does not retain memory of previous messages in the conversation.
+
+## Overview
+The main code utilizes the Microsoft Semantic Kernel to integrate with the OpenAI GPT-3.5-turbo model. Here's a breakdown of what the code does:
```
Comment (stephentoub) [inline src/build-2024/01 - Hello Semantic Kernel/README.md:4]:
```suggestion
The main code utilizes the Microsoft Semantic Kernel library to integrate with the OpenAI GPT-3.5-turbo model. Here's a breakdown of what the code does:
```

File: src/build-2024/02 - Add Chat History/Program.cs [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+    .Build();
+
+var chatService = kernel.GetRequiredService<IChatCompletionService>();
+ChatHistory chatHistory = [];
+
+// Basic chat
+while (true)
+{
+    Console.Write("Q: ");
+    chatHistory.AddUserMessage(Console.ReadLine());// Add user message to chat history, then it can be use to get more context for the next chat response
```
Comment (stephentoub) [inline src/build-2024/02 - Add Chat History/Program.cs:15]:
```suggestion
    chatHistory.AddUserMessage(Console.ReadLine()); // Add user message to chat history
```

File: src/build-2024/02 - Add Chat History/Program.cs [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+    var response = await chatService.GetChatMessageContentsAsync(chatHistory);// Get chat response based on chat history
```
Comment (stephentoub) [inline src/build-2024/02 - Add Chat History/Program.cs:16]:
```suggestion
    var response = await chatService.GetChatMessageContentsAsync(chatHistory); // Get chat response based on chat history
```

File: src/build-2024/02 - Add Chat History/Program.cs [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+    Console.WriteLine(response[response.Count - 1]);
```
Comment (stephentoub) [inline src/build-2024/02 - Add Chat History/Program.cs:17]:
```suggestion
    Console.WriteLine(response); // print out the response
```

File: src/build-2024/02 - Add Chat History/Program.cs [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+    chatHistory.AddRange(response);// Add chat response to chat history, hence it can be use to get more context for the next chat response
```
Comment (stephentoub) [inline src/build-2024/02 - Add Chat History/Program.cs:18]:
```suggestion
    chatHistory.AddRange(response); // Add chat response to chat history
```

File: src/build-2024/02 - Add Chat History/README.md [hunk 7]
```diff
@@ -0,0 +1,10 @@
+﻿## Console Chat Application with Stateful Chat AI
+
+This console application demonstrates a basic chat interface using OpenAI's GPT model (such as GPT-3.5-Turbo or GPT-4-Turbo) through the Microsoft Semantic Kernel. The application maintains a stateful chat interface, where it continuously takes user input and provides responses generated by the AI model, while remembering previous contexts.
+### Chat Service
+`kernel.GetRequiredService<IChatCompletionService>()` retrieves the chat completion service from the kernel.
+
+#### Chat History
+Named `chatHistory` is used to store the conversation history. Each user message and AI response is appended to this list, maintaining the context of the conversation.
+
+By maintaining a chat history, the application ensures that the AI model can remember previous contexts and provide more coherent and contextually relevant responses.
```
Comment (stephentoub) [inline src/build-2024/02 - Add Chat History/README.md:10]: I assume this is a still a work in progress :)

File: src/build-2024/05 - Add Plugin (Bing Search)/Program.cs [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+
+// Add logging services to the builder
+// builder.Services.AddLogging(b => b.AddConsole().SetMinimumLevel(LogLevel.Trace));
+
+var kernel = builder
+    .AddOpenAIChatCompletion(openAIChatCompletionModelName, Environment.GetEnvironmentVariable("OPENAI_API_KEY")) // add the OpenAI chat completion service.
+    .Build();
+
+#pragma warning disable
+kernel.ImportPluginFromObject(new Microsoft.SemanticKernel.Plugins.Web.WebSearchEnginePlugin(
```
Comment (stephentoub) [inline src/build-2024/05 - Add Plugin (Bing Search)/Program.cs:20]:
```suggestion
kernel.ImportPluginFromObject(new WebSearchEnginePlugin(
```

File: src/build-2024/05 - Add Plugin (Bing Search)/Program.cs [hunk 9]
```diff
@@ -0,0 +1,38 @@
+﻿using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Logging;
+using Microsoft.SemanticKernel;
+using Microsoft.SemanticKernel.ChatCompletion;
+using Microsoft.SemanticKernel.Connectors.OpenAI;
+using Microsoft.SemanticKernel.Plugins.Web.Bing;
```
Comment (stephentoub) [inline src/build-2024/05 - Add Plugin (Bing Search)/Program.cs:6]:
```suggestion
using Microsoft.SemanticKernel.Plugins.Web;
using Microsoft.SemanticKernel.Plugins.Web.Bing;
```

File: src/build-2024/05 - Add Plugin (Bing Search)/Program.cs [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+
+var openAIChatCompletionModelName = "gpt-4-turbo"; // this could be other models like "gpt-4-turbo".
```
Comment (stephentoub) [inline src/build-2024/05 - Add Plugin (Bing Search)/Program.cs:8]:
```suggestion
var openAIChatCompletionModelName = "gpt-4-turbo"; // this could be other models like "gpt-4o".
```

File: src/build-2024/06 - Modifying Kernel Behavior with Dependency Injection/Program.cs [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+
+// injecting services to the kernel such as logging, http client, redaction.
+// builder.Services.AddLogging(b => b.AddConsole().SetMinimumLevel(LogLevel.Trace));
+
+builder.Services.ConfigureHttpClientDefaults(b =>
+{
+    b.AddStandardResilienceHandler();
+    b.RedactLoggedHeaders(["Authorization"]);
+});
+builder.Services.AddRedaction();
```
Comment (stephentoub) [inline src/build-2024/06 - Modifying Kernel Behavior with Dependency Injection/Program.cs:21]: This redaction stuff is only relevant if the AddLogging above is uncommented.

File: src/build-2024/07 - Using Semantic Kernel in WebApp/Program.cs [hunk 12]
```diff
@@ -0,0 +1,28 @@
...
+var app = builder.Build();
+
+app.MapGet("/WeatherForecast", async (Kernel kerel) =>
+{
+    int temp = Random.Shared.Next(-20, 55);
+    return new WeatherForecast
+    (
+        DateOnly.FromDateTime(DateTime.Now),
+        temp,
+        await kerel.InvokePromptAsync<string>($"Short description of weather at {temp}0c?") // This description will be generated by the AI model for the given temperature.
```
Comment (stephentoub) [inline src/build-2024/07 - Using Semantic Kernel in WebApp/Program.cs:20]:
```suggestion
        await kerel.InvokePromptAsync<string>($"Short description of weather at {temp} degrees Celsius") // This description will be generated by the AI model for the given temperature.
```

File: src/build-2024/06 - Modifying Kernel Behavior with Dependency Injection/Program.cs (hunk 11 above)
Comment (zewditu) [reply src/build-2024/06 - Modifying Kernel Behavior with Dependency Injection/Program.cs:21]: @JakeRadMSFT  hence  we do not need a separate "Add Logging" project, we can merge it here

File: src/build-2024/02 - Add Chat History/Program.cs (hunk 5 above)
Comment (zewditu) [reply src/build-2024/02 - Add Chat History/Program.cs]: @stephentoub  response is not a simple value rather it is ` IReadOnlyList<ChatMessageConetent>`

Comment (zewditu) [reply src/build-2024/02 - Add Chat History/Program.cs]: NVM the method we have to call should be `GetChatMessageContentAsync` instead of `GetChatMessageContentsAsync`

File: src/build-2024/02 - Add Chat History/README.md (hunk 7 above)
Comment (zewditu) [reply src/build-2024/02 - Add Chat History/README.md:10]: yes it is

File: src/build-2024/02 - Add Chat History/README.md [hunk 13]
```diff
@@ -0,0 +1,53 @@
+﻿# Exercise : Add Chat History
+
+This console application demonstrates a basic chat interface using OpenAI's GPT model (such as GPT-3.5-Turbo or GPT-4-Turbo) through the Microsoft Semantic Kernel. The application maintains a stateful chat interface, where it continuously takes user input and provides responses generated by the AI model, while remembering previous contexts.
+
+## Prerequisites
+
+ You can refer [here](https://github.com/zewditu/ai-samples/blob/U/zewd/Build2024/src/build-2024/01%20-%20Hello%20Semantic%20Kernel/README.md#prerequisites)
```
Comment (JakeRadMSFT) [inline src/build-2024/02 - Add Chat History/README.md:7]: Use relative path

File: src/build-2024/04 - Add Logging/README.md [hunk 14]
```diff
@@ -0,0 +1,33 @@
+﻿# Exercise - Add Logging
+
+## Overview
+
+> Most of the code in this project is the same with `02 - Add Chat History` project codes. You can refer its [README.md](../02%20-%20Add%20Chat%20History/README.md). In this readme only specifc codes for this project are explained.
+
```
Comment (JakeRadMSFT) [inline src/build-2024/04 - Add Logging/README.md:6]: Can you add a section for requred nuget packages also?

Comment (JakeRadMSFT) [reply src/build-2024/04 - Add Logging/README.md:6]: For all the readme

File: src/build-2024/docs/Exercise/01 Hello Semantic Kernel.md [hunk 15]
```diff
@@ -0,0 +1,76 @@
...
+<!--TODo: Time to complelet the exercise should be included-->
+Now we have an understanding of semantic kernel library and  chat completions,let's create a basic console application that uses them.
+
+## Create the console application
+
+1. Run the following command on `PowerShell` to create a new .NET application named **01 - Hello Semantic 
+  Kernel**.
+
+    ```shell
+    dotnet new console -n 01 - Hello Semantic Kernel
```
Comment (JakeRadMSFT) [inline src/build-2024/docs/Exercise/01 Hello Semantic Kernel.md:11]: HelloBuild

File: src/build-2024/docs/Exercise/07 Using Semantic Kernel in WebApp.md [hunk 16]
```diff
@@ -0,0 +1,93 @@
+﻿# Exercise - Using Semantic Kernel in WebApp
+
+## Create the web application
+
+1. Run the following command on `PowerShell` to create a new .NET  web application named **07 - Using Semantic Kernel in WebApp**.
+
+    ```shell
+    dotnet new webapp -n 07 - Using Semantic Kernel in WebApp
```
Comment (JakeRadMSFT) [inline src/build-2024/docs/Exercise/07 Using Semantic Kernel in WebApp.md:8]: HelloBuildMinimalWebApis

File: src/build-2024/docs/Exercise/01 Hello Semantic Kernel.md (hunk 15 above)
Comment (JakeRadMSFT) [inline src/build-2024/docs/Exercise/01 Hello Semantic Kernel.md:11]:
```suggestion
    dotnet new console -n 01 - HelloBuild
```

File: src/build-2024/docs/Exercise/07 Using Semantic Kernel in WebApp.md (hunk 16 above)
Comment (JakeRadMSFT) [inline src/build-2024/docs/Exercise/07 Using Semantic Kernel in WebApp.md:8]:
```suggestion
    dotnet new webapp -n HelloBuildMinimalWebApis
```

File: src/build-2024/docs/Introduction.md [hunk 17]
```diff
@@ -0,0 +1,5 @@
+
```
Comment (jmatthiesen) [inline src/build-2024/docs/Introduction.md:1]: I assume this is still a work in-progress along with the other overview articles.

Be sure to add a step in one of these to set the environment variable that will be used for the samples.

File: src/build-2024/docs/Exercise/01 Hello Semantic Kernel.md [hunk 18]
```diff
@@ -0,0 +1,79 @@
...
+    while (true)
+    {
+        Console.Write("Q: ");
+        Console.WriteLine(await kernel.InvokePromptAsync(Console.ReadLine()!));
+    }
+    ```
+
+    - This loop continuously prompts the user for input, sends the input to the OpenAI model, and prints the AI's response.
+
+1. Let's see what we have so far, you can run the application by entering `dotnet run` into the terminal. Experiment with a user prompt "Hi my name is Alice" and a follow-up question "what is my name?" you will get something similar output as shown below on console.
```
Comment (jmatthiesen) [inline src/build-2024/docs/Exercise/01 Hello Semantic Kernel.md:59]:
```suggestion
1. Let's see what we have so far, you can run the application by entering `dotnet run` into the terminal. Experiment with a user prompt "Hi my name is Alice" and a follow-up question "what is my name?" your output may vary, but it will be similar to what is shown below.
```

File: src/build-2024/docs/Exercise/01 Hello Semantic Kernel.md [hunk 19]
```diff
@@ -0,0 +1,79 @@
...
+      Q: Hi my name is Alice
+      Hello Alice, pleased to meet you! How can I assist you today?
+      Q: What is my name?
+      I'm sorry, I cannot provide your name as I do not have that information. If you would like me to refer to you by a specific name during our conversation, please let me know.
+      Q:
+      ```
+
+## Complete sample project
+
+[01 Hello Semantic Kernel](../../01%20-%20Hello%20Semantic%20Kernel).
```
Comment (jmatthiesen) [inline src/build-2024/docs/Exercise/01 Hello Semantic Kernel.md:71]:
```suggestion
You can view the completed project in the [01 Hello Semantic Kernel](../../01%20-%20Hello%20Semantic%20Kernel) folder.
```

File: src/build-2024/docs/Exercise/02 Add Chat History.md [hunk 20]
```diff
@@ -0,0 +1,58 @@
...
+1. Add the following using statment at the top of `Program.cs` file.
+
+    ```csharp
+    using Microsoft.SemanticKernel.ChatCompletion;
+    ```
+
+1. Add the following code next to kernel initialization to get the required service from the kernel which is chat the completion service.
+
+    ```csharp
+    var chatService = kernel.GetRequiredService<IChatCompletionService>();
```
Comment (jmatthiesen) [inline src/build-2024/docs/Exercise/02 Add Chat History.md:16]: To make it easier to follow, consider showing the other code before this line. Perhaps add text that says

```suggestion
Change the code from 
    ```csharp
    var kernel = Kernel.CreateBuilder()
          .AddOpenAIChatCompletion(openAIChatCompletionModelName, Environment.GetEnvironmentVariable("OPENAI_API_KEY"))
          .Build();
    ```
to

    ```csharp
    var kernel = Kernel.CreateBuilder()
          .AddOpenAIChatCompletion(openAIChatCompletionModelName, Environment.GetEnvironmentVariable("OPENAI_API_KEY"))
          .Build();

    var chatService = kernel.GetRequiredService<IChatCompletionService>();
    ```
```

Or something like that? It does get a bit wordy, though. You could also try text like "After the line that starts with `var kernel...` add the following code. Maybe play around with that a little and see what you think?

File: src/build-2024/docs/Exercise/02 Add Chat History.md [hunk 21]
```diff
@@ -0,0 +1,58 @@
...
+    Q: Hi my name is Alice
+    Hello Alice! Nice to meet you. How can I assist you today?
+    Q: What is my name?
+    Your name is Alice.
+    Q:
+    ```
+
+## Complete sample project
+
+  [02 - Add Chat History](../../02%20-%20Add%20Chat%20History/) project.
```
Comment (jmatthiesen) [inline src/build-2024/docs/Exercise/02 Add Chat History.md:51]:
```suggestion
View the completed sample in the  [02 - Add Chat History](../../02%20-%20Add%20Chat%20History/) project.
```

File: src/build-2024/docs/Exercise/03 Add Plugin (Function Call).md [hunk 22]
```diff
@@ -0,0 +1,65 @@
...
+
+1. Add the following next to kernel initialization to import a plugin to the kernel from type `DemographicInfo`.
+
+    ```csharp
+    // Import the DemographicInfo class to the kernel, so it can be used in the chat completion service.
+    // this plugin could be from other options such as functions, prompts directory, etc.
+    kernel.ImportPluginFromType<DemographicInfo>();
+    ```
+
+1. Add the following function calling  behavior setting. The setting is configured to call the method GetAge automatically when the user requests the age of the person with the provided name.
```
Comment (jmatthiesen) [inline src/build-2024/docs/Exercise/03 Add Plugin (Function Call).md:39]:
```suggestion
1. Add the following function calling behavior setting. The setting is configured to call the method `GetAge` automatically when the user requests the age of the person with the provided name.
```

File: src/build-2024/docs/Exercise/03 Add Plugin (Function Call).md [hunk 23]
```diff
@@ -0,0 +1,65 @@
...
+
+    ```console
+    Q: My name is Alice. How old am I?
+    Alice, you are 25 years old.
+    Q: 
+    ```
+
+## Complete sample project
+
+[03 Add Plugin (Function Call)](../../03%20-%20Add%20Plugin%20(Function%20Call))
```
Comment (jmatthiesen) [inline src/build-2024/docs/Exercise/03 Add Plugin (Function Call).md:61]:
```suggestion
View the completed sample in the [03 Add Plugin (Function Call)](../../03%20-%20Add%20Plugin%20(Function%20Call)) project.
```

File: src/build-2024/docs/Exercise/04 Add Logging.md [hunk 24]
```diff
@@ -0,0 +1,51 @@
...
+      Q: Hello
+      trce: Microsoft.SemanticKernel.Connectors.OpenAI.OpenAIChatCompletionService[0]
+            ChatHistory: [{"Role":{"Label":"user"},"Items":[{"$type":"TextContent","Text":"Hello"}]}], Settings: {"temperature":1,"top_p":1,"presence_penalty":0,"frequency_penalty":0,"max_tokens":null,"stop_sequences":null,"results_per_prompt":1,"seed":null,"response_format":null,"chat_system_prompt":null,"token_selection_biases":null,"ToolCallBehavior":null,"User":null,"logprobs":null,"top_logprobs":null,"model_id":null}
+      info: Microsoft.SemanticKernel.Connectors.OpenAI.OpenAIChatCompletionService[0]
+            Prompt tokens: 8. Completion tokens: 9. Total tokens: 17.
+      Hello! How can I help you today?
+      Q:
+      ```
+
+> **Note:**  From the outout on console notice the log  information that provided detailed information about our model settings.
```
Comment (jmatthiesen) [inline src/build-2024/docs/Exercise/04 Add Logging.md:43]:
```suggestion
> **Note:**  From the output on the console, notice the log information that provided detailed information about our model settings.
```

File: src/build-2024/docs/Exercise/04 Add Logging.md [hunk 25]
```diff
@@ -0,0 +1,51 @@
...
+
+## Complete sample project
+
+[04 Add Logging)](../../04%20-%20Add%20Logging/)
```
Comment (jmatthiesen) [inline src/build-2024/docs/Exercise/04 Add Logging.md:47]:
```suggestion
View the completed sample in the [04 Add Logging)](../../04%20-%20Add%20Logging/) project.
```

File: src/build-2024/docs/Exercise/05 Add Plugin (Bing Search).md [hunk 26]
```diff
@@ -0,0 +1,62 @@
...
+    8. **GPT-4o**: General availability for using GPT-4o with Azure credits, aimed at aiding developers to build and deploy applications more efficiently.
+
+    9. **Rewind Feature**: Introduction of Rewind, a new feature to improve the search functionality on Windows PCs, making it as efficient as web searches.
+
+    These announcements span various areas from AI advancements and new hardware to enhanced developer tools and software capabilities.
+    ```
+
+## Complete sample project
+
+[05 Add Plugin (Bing Search)](../../05%20-%20Add%20Plugin%20(Bing%20Search))
```
Comment (jmatthiesen) [inline src/build-2024/docs/Exercise/05 Add Plugin (Bing Search).md:58]:
```suggestion
View the completed sample in the [05 Add Plugin (Bing Search)](../../05%20-%20Add%20Plugin%20(Bing%20Search)) project.
```

File: src/build-2024/docs/Exercise/06 Modifying Kernel Behavior with Dependency Injection.md [hunk 27]
```diff
@@ -0,0 +1,312 @@
...
+      ```
+
+1. Add the following using statment at the top of `Program.cs` file.
+
+      ```cshap
+      using Microsoft.Extensions.Http;
+      ```
+
+1. Add the following code after `builder.Services.AddLogging(b => b.AddConsole().SetMinimumLevel(LogLevel.Trace));`
+ which modifies the kernel behavior by injecting services into it and suppress warning.
```
Comment (jmatthiesen) [inline src/build-2024/docs/Exercise/06 Modifying Kernel Behavior with Dependency Injection.md:30]:
```suggestion
 which modifies the kernel behavior by injecting services into it. For this sample we're using experimental features, so we add a `#pragma` statement to suppress warnings about using them.
```

File: src/build-2024/docs/Exercise/06 Modifying Kernel Behavior with Dependency Injection.md [hunk 28]
```diff
@@ -0,0 +1,312 @@
...
+            Ending HttpMessageHandler cleanup cycle after 0.0045ms - processed: 0 items - remaining: 1 items
+      dbug: Microsoft.Extensions.Http.DefaultHttpClientFactory[100]
+            Starting HttpMessageHandler cleanup cycle with 1 items
+      dbug: Microsoft.Extensions.Http.DefaultHttpClientFactory[101]
+            Ending HttpMessageHandler cleanup cycle after 0.0036ms - processed: 0 items - remaining: 1 items
+      ```
+
+## Complete sample project
+
+[06 Modifying Kernel Behavior with Dependency Injection](../../06%20-%20Modifying%20Kernel%20Behavior%20with%20Dependency%20Injection/))
```
Comment (jmatthiesen) [inline src/build-2024/docs/Exercise/06 Modifying Kernel Behavior with Dependency Injection.md:308]:
```suggestion
View the completed sample in the [06 Modifying Kernel Behavior with Dependency Injection](../../06%20-%20Modifying%20Kernel%20Behavior%20with%20Dependency%20Injection/)) project.
```

File: src/build-2024/docs/Exercise/07 Using Semantic Kernel in WebApp.md [hunk 29]
```diff
@@ -0,0 +1,110 @@
...
+    {
+        "date": "2024-06-17",
+        "tempratureC": 4,
+        "summary": "At 4 degrees Celsius, the weather is cool and slightly chilly. It is above freezing, so there is no ice, but it's cold enough that you might want a jacket or sweater when outdoors. This temperature is typical for late autumn or early spring in temperate regions."
+    }
+    ```
+
+## Complete sample project
+
+[07 Using Semantic Kernel in WebApp](../../07%20-%20Using%20Semantic%20Kernel%20in%20WebApp/))
```
Comment (jmatthiesen) [inline src/build-2024/docs/Exercise/07 Using Semantic Kernel in WebApp.md:106]:
```suggestion
View the completed sample in the [07 Using Semantic Kernel in WebApp](../../07%20-%20Using%20Semantic%20Kernel%20in%20WebApp/)) project.
```

**Review by jmatthiesen — CHANGES_REQUESTED** [review]

File: src/build-2024/01 - Hello Semantic Kernel/README.md [hunk 30]
```diff
@@ -0,0 +1,3 @@
+# Hello semantic kernel
+
+A complete documentation for this project is [here](../docs/Exercise/01%20Hello%20Semantic%20Kernel.md)
```
[the remark this answers is not in the document]
Comment (JakeRadMSFT) [reply src/build-2024/01 - Hello Semantic Kernel/README.md:3]: These were added as feedback from Luis. He was looking for readme here.

[the remark this answers is not in the document]
Comment (zewditu) [reply src/build-2024/01 - Hello Semantic Kernel/README.md:3]: This is to make the readme files in exercise folder more discoverable
@JakeRadMSFT @luisquintanilla  thought

File: src/build-2024/02 - Add Chat History/02 - Add Chat History.csproj [hunk 31]
```diff
@@ -0,0 +1,11 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <OutputType>Exe</OutputType>
+    <TargetFramework>net8.0</TargetFramework>
+    <ImplicitUsings>enable</ImplicitUsings>
+  </PropertyGroup>
+  <ItemGroup>
+    <PackageReference Include="Microsoft.SemanticKernel" Version="1.13.0" />
```
[the remark this answers is not in the document]
Comment (zewditu) [reply src/build-2024/02 - Add Chat History/02 - Add Chat History.csproj:9]: 1.15.0 is out at this time

File: src/build-2024/04 - Add Logging/Program.cs [hunk 32]
```diff
@@ -0,0 +1,30 @@
...
+using Microsoft.Extensions.Logging;
+using Microsoft.SemanticKernel;
+using Microsoft.SemanticKernel.ChatCompletion;
+
+var openAIChatCompletionModelName = "gpt-3.5-turbo"; // this could be other models like "gpt-4o".
+
+var builder = Kernel.CreateBuilder();
+
+// Add logging services to the builder
+builder.Services.AddLogging(b => b.AddConsole().SetMinimumLevel(LogLevel.Trace));
```
[the remark this answers is not in the document]
Comment (zewditu) [reply src/build-2024/04 - Add Logging/Program.cs:11]: Each of these projects can run independently and do not need to be learned in sequence. Since the code is not essential for running the project, I don't believe it should be shared across projects.

File: src/build-2024/docs/Introduction.md (hunk 17 above)
Comment (zewditu) [reply src/build-2024/docs/Introduction.md]: It is listed [here](https://github.com/zewditu/ai-samples/blob/U/zewd/Build2024/src/build-2024/README.md) as one of the prerequisites.

**Review by jmatthiesen — APPROVED** [review]

**Review by luisquintanilla — APPROVED** [review]
@zewditu LGTM. We might want to update the version of SK at some point but don't see that as a blocker at this time.

**Comment by jmatthiesen** [comment]
This is merged now, thanks @zewditu !
